### ISSUE TITLE
Pathological worstcase when password is a hex string

### ISSUE THREAD

**Issue by Mjolnir42** [nbutton23/zxcvbn-go#15]
Hi,

I think I hit an algorithmic pathological worst case, if the supplied password string is a hex encoded string. I have not been able to reproduce this behavior with other, way longer passphrases.

I've created a gist with a testcase: [Gist](https://gist.github.com/Mjolnir42/fbbb9691f1d39e2b432797bbbbb4f3bb)

At the point where I aborted (password length 32 characters), it is was consuming around 7GBytes of RAM.

**Comment by Mjolnir42** [comment]
Since the initial data came from my laptop, I've also tried it on my workstation. It could not complete the 32 characters before it ran out of memory:
```
swap_pager: out of swap space
swap_pager_getswapspace(16): failed
pid 1403 (zxcvbn), uid 1001, was killed: out of swap space
```
I've updated the data comment in the gist, including top output from shortly before it ended. Memory consumption was around 14.5GByte for the 32 character string at that time.

**Comment by nbutton23** [comment]
hmmm.  Ill have to play with it and see if I can figure out where its spending the most time, and why its eating so much memory.

**Comment by nbutton23** [comment]
So I found where its spending the time. Im not sure if this is where the memory issue is. 

https://gist.github.com/nbutton23/b11f57561c6471890bd95ac5a3b52a75

**Comment by nbutton23** [comment]
Okay after some more digging. I have found that the issue is not in the above code. using pprof I can see that it is in matching.leet.go:getAllPermutationsOfLeetSubstitutions.    This is generating an insane number of permutations.

**Comment by t3h2mas** [comment]
I ran into this when running against a MD5 hash. Is there a workaround or an option to skip the leet permutations?

**Comment by nbutton23** [comment]
So @TheMacies just added the ability to filter out matchers.  

the call would look like `zxcvbn.PasswordStrength(password, nil, matching.FilterL33tMatcher)`

**Comment by cerlestes** [comment]
I ran into the same issue with a similiar string. For testing purposes, I've used UUIDs as passwords, like so:

```
zxcvbn.PasswordStrength("ce59f9ce-c1ca-4929-ad94-a494c7fbe032", nil)
```

After a few minutes I've noticed I still had no response from the API only to find that it was consuming nearly 11 GByte of RAM.

This issue needs to be addressed, as it allows for pretty trivial DoS attacks against any software using this library.

**Comment by TheMacies** [comment]
As a temporary solution, try out the feature I added a while ago.
https://github.com/nbutton23/zxcvbn-go/pull/24

**Comment by TheMacies** [comment]
Yes. Although i did not want to change the behaviour of this lib after someone updates  vendor. That was my primary goal

**Comment by cerlestes** [comment]
@edwincarlo Very nice work, thank you. Code is looking good. I'll give the patch a try next week and report back.

**Comment by sahib** [comment]
@edwincarlo: Great work, I see definite improvements. For a hash string like ``68b329da9893e34099c7d8ad5cb9c940`` the time has gone from ~7s to a few milliseconds. This finally allows using this library for checking passwords while typing without any crude workarounds.

For reference, the pprof output I made while testing the change:
[Before](https://i.imgur.com/i1nx4ON.png) applying @edwincarlo's changes and [after](https://i.imgur.com/gOEZbBN.png).

**Comment by cerlestes** [comment]
Awesome! Instead of killing the process after 60 seconds and 14GB of RAM consumed, my tests are now running through in ~0.1s, when the patch was applied (taken from current master branch).

@edwincarlo Thanks again!
@nbutton23 Please release asap :) Thanks!

**Comment by dcormier** [comment]
Is this still an issue? The last couple of comments suggest that it's been resolved.
